### DISCUSSION TITLE
Fluidsynth and the soundfont effects send problem

### DISCUSSION THREAD

**Discussion by spessasus** [FluidSynth/fluidsynth#1418] [Ideas]
# The problem
As you might know, the soundfont standard defines modulators for reverb and chorus MIDI controllers. Unfortunately, that amount is very small: only 20% of the full effect at the controller value of 127.

This causes chorus and reverb to be barely audible even in MIDIs that max it out. This affects soundfonts that don't bother to change these, i.e. most of them.

# Potential solutions

## 1. Stay as is
Currently, fluidsynth implements the said behavior of 20% by default and leaves it up to the soundfont to change.

### pros
- stays true to the spec
- can be modified by the soundfont 

### cons
- many soundfonts lack modulators at all, being just samples mapped to channels, so they have weak effects

## 2. Increase effect gain
This approach increases the effects' gain amount, by for example 1.8 to make the 20% be as loud as 100%.

### pros
- effects are now audible with default modulators 

### cons
- soundfonts that override the amount with larger values (like GUGS) will have too loud effects

## 3. BASSMIDI approach
Reading the BASS forums, BASSMIDI defaults to 100% modulators and they cannot be overriden.

### pros
- always audible effects, no matter if the soundfont has modulators or not

### cons
- effect modulators cannot be disabled
- effects amount cannot be changed

## 4. Spessasynth approach
I've addressed this problem differently:
- the default modulators are left at 200
- all modulators identical to the default effect modulators have their amount multiplied by 5 if their amount is below 1000, then clamped to 1000. If the transform amount is greater or equal to 1000, they are unchanged

### pros
- effects are always audible
- modulators still can be overriden and disabled
- effects amount can be lowered from 200 and will be scaled from 0-200 to 0-1000 respectively
- since the modulators can be disabled, effects can have a different curve

### cons
- effect modulators with the amount of 201 to 1000 will be changed to 1000

### which one to choose?

Personally, think that the problem of 99% of soundfonts lacking modulators should be addressed. I propose my solution (number 4), but it could be an option (like `synth.full-cc-effects`)

What do you all think? What's the best solution?

**Comment by derselbst** [comment]
So far I never had any problems with the effects being too "decent". I'd tend to keep it as is. Maybe @mrbumpy409 has an opinion about this as well?

**Comment by mrbumpy409** [comment]
I think FluidSynth should continue to honor the spec. If someone wants more reverb / chorus, they can A) modify the SoundFont by overriding the default modulators, as I have done with GeneralUser GS, or B) increase FluidSynth's reverb/chorus levels.

**Reply by spessasus** [reply]
>  A) modify the SoundFont by overriding the default modulators, as I have done with GeneralUser GS

That's the problem. Many (if not most) of soundfonts out there simply do not bother to set up modulators.
Citing from your *Using SoundFonts in 2016* blog post I've read a while back:

> Most software SoundFont synths are not compatible with the 2.01 spec, and many aren’t compatible with any version of the spec! **This has lead many SoundFont designers to code for the lowest common denominator. Alas, many SoundFonts today are nothing more than samples mapped to keys, with scant little in the way of actual sound programming.**

So that's why it's a problem IMO. GeneralUserGS is so special because it actually uses modulators and SF programming. Most other banks are just samples  + key and velocity ranges. This means very quiet effects on most of them.

> or B) increase FluidSynth's reverb/chorus levels.

I addressed this solution's pros and cons in my initial post:

> soundfonts that override the amount with larger values (like GUGS) will have too loud effects

If I tell fluidsynth to increase the effects gain, then soundfonts without modulators will sound good, but soundfonts that use the full range of effects send (like yours) will simply be too loud.

So, if I'm using fluidsynth in my own software, there's nothing I can do to make both kinds of soundfonts have rougly equal effects. Increasing gain will break GUGS. Leaving it as-is will cause GUGS to sound fine, but the *modulator-less* sf's will have quiet effects.

Actually, while making spessasynth, I forgot that the send was 200 by default. So I was wondering why the effects were so quiet. I kept increasing their gain until they were audible enough (i was testing with a bank with almost no modulators). Then I tried using GUGS and the effects were so loud that it hurt, since it was using 1000 send instead of 200 which I adjusted the gain for. That's why I brought up this in the first place and implemented a solution I described above.

**Reply by mrbumpy409** [reply]
> So, if I'm using fluidsynth in my own software, there's nothing I can do to make both kinds of soundfonts have rougly equal effects. Increasing gain will break GUGS. Leaving it as-is will cause GUGS to sound fine, but the _modulator-less_ sf's will have quiet effects.

I can see your point. I'll have to think more about your "Spessasynth approach" idea. At the very least, perhaps it could be a toggleable feature. One thing I'm hoping to be able to study is the reverb & chorus send behavior of the Sound Blaster Live! and AWE32/64 cards. It seems that pre-SoundFont 2.01, the FX send via CC91 and CC93 was more closely matched to the sensitivity of the reverb and chorus generators set in the SoundFont's instruments/presets. I just need somebody with one of these cards to run my [SoundFont spec test](https://github.com/mrbumpy409/SoundFont-Spec-Test) and post the recording [here](https://github.com/mrbumpy409/SoundFont-Spec-Test/issues/2). Once I can determine what the sensitivity level used to be, perhaps an optional feature could be added to FluidSynth to emulate SoundFont 2.0 and earlier effects send behavior, which would provide a much better effects balance when playing old AWE32 songs, and also remedy the issue you are reporting here.

> Actually, while making spessasynth, I forgot that the send was 200 by default. So I was wondering why the effects were so quiet. I kept increasing their gain until they were audible enough (i was testing with a bank with almost no modulators). Then I tried using GUGS and the effects were so loud that it hurt, since it was using 1000 send instead of 200 which I adjusted the gain for. That's why I brought up this in the first place and implemented a solution I described above.

To be fair, my FX send levels of 1000 in the older GeneralUser GS were intended for use with my custom Audigy EAX presets. The current GeneralUser GS version doesn't increase the fx send levels nearly so much. On most instruments, reverb send is 350 and chorus send is 300. I also use around 0-60 reverb send on kick drums, similar to what Roland does, since reverb on kick drums pretty much always sounds bad. These values were tailored around FluidSynth 2.4's new default reverb and chorus settings, but also sound good with my current Audigy2 ZS EAX preset.

**Comment by mrbumpy409** [comment]
Thanks to user @Kuokka77, I was able to measure the AWE64 effects response for internal vs. CC91/93 values (which should also be the same for Sound Blaster Live!, but I will need a recording with FX on that hardware to confirm). From this test, I can ascertain that the CC91/93 modulation of reverb/chorus amount is 400 (40%), compared to the SoundFont 2.01/2.04 default of 200 (20%).

Furthermore, this test allowed me to attempt to balance FluidSynth's effects levels and settings to better match the AWE32/64's "Hall 1" and "Chorus 1" presets. With the SoundFont configured for 40% CC91/93 FX send, the following FluidSynth effects settings should result in a decent effects level correspondence to the AWE32/64:
```
set synth.reverb.damp 0.3
set synth.reverb.level 0.6
set synth.reverb.room-size 0.65
set synth.reverb.width 0.8
set synth.chorus.depth 4.7
set synth.chorus.level 0.43
set synth.chorus.nr 3
set synth.chorus.speed 0.4
```
I have included more information and recordings in [my comment here](https://github.com/FluidSynth/fluidsynth/pull/1434#issuecomment-2529352355).
